src/js/helpers/DataPanel.ts

```diff
@@ -0,0 +1,137 @@
+  for await (const sublayer of sublayersArray) {
+    const url = sublayer.url;
+    const qParams = {
+      where: '1=1',
+      outFields: ['*'],
+      geometry: geometry,
+      returnGeometry: false
+    };
+    try {
+      const sublayerResult = await esriQuery(url, qParams);
+      //if no features are found, skip it
+      if (sublayerResult.features.length !== 0) {
+        const features = sublayerResult.features.map(f => {
+          return {
+            attributes: f.attributes,
+            geometry: f.geometry
+          };
+        });
+        const subCleanedResult = {
+          layerID: sublayer.layer.id,
+          layerTitle: sublayer.layer.title,
+          features: features
+        };
+        processedSubsResults.push(subCleanedResult);
+      }
+    } catch (e) {
+      console.log(e);
+    }
+  }
+  return processedSubsResults;
+}
+
+async function fetchAsyncServerResults(
+  map: Map | undefined,
+  mapview: MapView,
+  mapPoint: Point
+): Promise<any> {
+  if (map) {
+    //Iterate over map layers, filter out turned off layers and non map-image layers
+    const visibleServerLayers = map.layers.filter(
+      l => l.visible && l.type === 'map-image'
+    );
+    //Extract all sublayers
+    const sublayers: any = visibleServerLayers
+      .flatten((item: any) => item.sublayers)
+      .filter((l: any) => !l.sublayers);
+
+    //We need to convert Collection of sublayers to regular array for async processing
+    const sublayersArray: any[] = sublayers.items.map((sl: Sublayer) => sl);
+
+    //TODO: Better way to handle mouse click buffering?
+    const distance = mapview.resolution * 0.005;
+    const geometry = geometryEngine.buffer(mapPoint, distance, 'miles');
+    const processedSubs: LayerFeatureResult[] = await processSublayers(
+      geometry,
+      sublayersArray
+    );
+    return processedSubs;
+  }
+}
+
+//Client Side Feature Fetching
+export function addPopupWatchUtils(
+  mapview: MapView,
+  map: Map | undefined,
+  mapPoint: Point
+): void {
+  //Watching for promises array on Popup, that gets resolved once user clicks on the map and features are returned. Note: this will not return any server side features.
+  const popupWatcher = watch(mapview.popup, 'promises', promises => {
+    function resolveClientPromisesWithErrors(promises: any): Promise<void> {
+      return Promise.all<any>(
+        promises.map((p: Promise<any>) => p.catch((error: Error) => null))
+      ).then(async clientFeatures => {
+        let layerFeatureResults: LayerFeatureResult[] = [];
+        const cleanClientFeautures = clientFeatures
+          .filter(f => f !== null) //catch failed promises
+          .filter(f => f.length !== 0) //catch no features returned
+          .map((featureObject: Graphic[]) => {
+            const newFeatureObject: LayerFeatureResult = {
+              layerID: featureObject[0].layer.id,
+              layerTitle: featureObject[0].layer.title,
```

Comment:
Is `this` the place to add a PopupTemplate for each layer? Should that be in `allAvailableLayers` in State?